Introduce visibility toggle on the outline panel

sprocketc opened the pull request:
Introduced a visibility toggle to the outline tree. The toggle will be visible on mouse over, or when the node is hidden.

![visibility-toggle](https://github.com/user-attachments/assets/3a4b91f9-d970-4f74-bdfb-2f0405aaa439)

Fixes #6026

### Technical changes
Rendering an additional button on the tree cell that sticks to the right side of the panel, without blocking horizontal scrolling.

sprocketc committed:
wip - visibility toggle

sprocketc committed:
add icons and fix various issues

sprocketc committed:
optimize TreeCell to avoid flickering

sprocketc committed:
fix various minor issues

sprocketc committed:
enhance styles

sprocketc committed:
fix naming

sprocketc committed:
Merge branch 'dev' into issue-3916-visibility-toggle

sprocketc committed:
fix hideable property

sprocketc committed:
fix visibility button flickering

sprocketc committed:
toggle visibility on individual nodes

sprocketc committed:
refactor

sprocketc committed:
Merge branch 'dev' into issue-3916-visibility-toggle

sprocketc committed:
update tilemap layer outline on visibility change

sprocketc commented on editor/src/clj/editor/outline_view.clj:
```diff
@@ -112,23 +115,24 @@
                 :children (mapv :item data)
                 :child-error? (boolean (some :child-error? data))
                 :child-overridden? (boolean (some :child-overridden? data))
+                :hideable? (contains? outline-name-paths (rest node-outline-key-path))
```
Is there an easier way to get to this info? Connecting to `outline-name-paths` seemed like the logical solution here, given the current implementation.

matgis replied:
I think both `hideable` and `hidable` are alternative but valid ways to spell it?

Regarding using `outline-name-paths`, it works for now, but may not be exactly what we want. For the existing "hide this and everything below it" action, we want to be able to select a "group" element like for example a Game Object, and hide everything related to it. But the Game Object itself doesn't have a visual component - it is the various components below it that may or may not be hideable. So technically I think we may want to not show the eye icon next to Game Objects, just like we do not show it for things like Script components.

We might be able to check if the `child-scene` contains a `:renderable` when we produce the `outline-name-paths` in `scene-visibility/scene-outline-name-paths`, but I think we want the existing behavior for the "hide this and everything below it" action, so maybe we need to restructure the return value to include a bool signifying whether or not the `outline-name-path` is associated with a `:renderable`.

I think we can do these changes in a separate PR, though.

sprocketc replied:
I will probably change the behavior to toggle all children, so this case would make more sense.

sprocketc commented on editor/src/clj/editor/outline_view.clj:
```diff
@@ -458,9 +463,29 @@
   [^TreeView tree-view drag-entered-handler drag-exited-handler]
-  (let [cell (proxy [TreeCell] []
+  (let [eye-icon-open (app-view/make-svg-icon-graphic eye-open-path)
+        eye-icon-closed (app-view/make-svg-icon-graphic eye-closed-path)
```
Part of the initialization was moved here for performance purposes. That solved almost all blinking issues on visibility toggling, but there are still some rare cases where a tree cell might appear to be re-rendered, although it shouldn't. Is there anything else we can do?

matgis replied:
I suspect there might be a bug in the JavaFX redraw code. I've certainly seen issues around hover highlighting and the disclosure triangle earlier. I don't think we should worry about it for this PR.

sprocketc commented on editor/src/clj/editor/scene_visibility.clj:
```diff
@@ -254,6 +255,16 @@
              (hide-outline-name-paths! scene-visibility (g/node-value scene-visibility :selected-hideable-outline-name-paths))
              (show-outline-name-paths! scene-visibility (g/node-value scene-visibility :selected-showable-outline-name-paths)))))))
 
+(handler/defhandler :hide-toggle :workbench
+  (active? [scene-visibility evaluation-context user-data]
+           (g/node-value scene-visibility :active-scene-resource-node evaluation-context))
+  (run [scene-visibility user-data]
+       (let [{:keys [node-outline-key-path]} user-data
+             outline-name-paths #{(vec (rest node-outline-key-path))}]
+         (if (contains? (g/node-value scene-visibility :hidden-node-outline-key-paths) node-outline-key-path)
+           (show-outline-name-paths! scene-visibility outline-name-paths)
+           (hide-outline-name-paths! scene-visibility outline-name-paths)))))
+
```
A new handler was introduced to toggle the visibility state of a single nodes. The existing context menu action is applied to all selected elements and its children. That might result in unexpected behavior, so I decided to deviate. For instance, if we try to unhide a node that has at least one visible children using the context menu, the child will turn invisible, instead of changing the visibility of the selected parent.

sprocketc commented on editor/src/clj/editor/outline_view.clj:
```diff
@@ -471,15 +496,20 @@
+                       (.setImage image-view-icon (icons/get-image icon))
+                       (.setText text-label text)
+                       (.setGraphic visibility-button visibility-icon)
+                       (proxy-super setGraphic pane)
```
I initially tried switching to a `TreeTableView` component that seemed like a better fit, but that would require a lot of changes that might produce more problems down the line. I worked around the issue here by using `setGraphic` to render the icon+label+toggle.

sprocketc commented on editor/src/clj/editor/outline_view.clj:
```diff
@@ -458,9 +463,29 @@
 (defn make-tree-cell
   [^TreeView tree-view drag-entered-handler drag-exited-handler]
-  (let [cell (proxy [TreeCell] []
+  (let [eye-icon-open (app-view/make-svg-icon-graphic eye-open-path)
```
Not that important, but should we move `make-svg-icon-graphic` under `editor.ui.clj`, or maybe `editor.icons.clj`?

sprocketc commented on editor/styling/stylesheets/modules/_outline.scss:
```diff
@@ -76,6 +68,39 @@
       @include colorize-disclosure-arrow(-df-error-severity-fatal-dim);
     }
 
+    .label {
+      -fx-min-width: 40px;
+    }
```
This PR enables ellipsis on text-overrun, to allow displaying the toggle at the right side of the cell. That might not be ideal on deeply nested outlines. Although horizontal scrolling is not ideal either, this is a middle ground solution that always shows part of the label, and enables horizontal scrolling sooner. For what is worth, I tried to make the visibility toggle stick to the right, without blocking scroll, but I didn't manage to produce a good result. The main problem was that it required watching the scrolling position to rerender the button on a different anchor point, and I couldn't make this work properly.

matgis replied:
In many ways it would be nicer if the eye icons were in their own column to the right of the tree view item, but I understand that would necessitate more drastic changes to the view.

For now, I think this is fine. Maybe we'll revisit this later if we need to!

matgis commented on editor/src/clj/editor/tile_map.clj:
```diff
@@ -486,7 +486,7 @@
 
 (g/defnk produce-layer-outline
-  [_node-id id z]
+  [_node-id id z visible]
```
The same is true for Gui scene nodes, BTW. They have a Visible checkbox.

With the previous "hide this and everything below it" functionality, we explicitly didn't want to tie it to the setting that affects visibility in-game - it was considered more of a "scene management" feature for users of the editor, so they could get rid of clutter and focus on particular parts of the scene.

With the eye-icon visibility toggle, I personally think it should work the same way (and it nicely does interact with the existing show / hide stack 👍). However, I'm sure some users might expect it to be tied to the in-game state as well.

For now, I think we should leave it as implemented, and if enough people yell about it, we might reconsider. 😅

sprocketc replied:
I don't think we should tie those options. What I found weird was that if you make the layer invisible, you are blocking the in-editor visibility toggling. For example, if I turn the in-game visibility of a tilemap layer off, and then hide the parent, the change won't be propagated to the layer. Now that I think about it, it will be weird if the state on the view is not synced with the icon, so I guess it's fine.

matgis commented on editor/resources/scene/images/eye_closed.svg:
Beautiful icons! ❤️ Did you make these?

sprocketc replied:
They are based on the existing `eye_icon_eye_arrow.svg` for consistency.

matgis reviewed (dismissed):
Works great! ❤️

vlaaad commented:
Before I got sick, I started reviewing it and wanted to discuss aspects of this PR:
- are we sure that using ellipsis instead of horizontal scroll bar is not a problem? In gui files with lots of nested templates, a lot of nodes in the outline view share the same (very long) text prefix. It might be a problem
- should the on/off switch be recursive? I.e. should turning off go visibility make its sprites invisible? Currently it does not work like that, and I found it puzzling in my tests 

Either way, the improvements are very cool!

sprocketc commented:
> * are we sure that using ellipsis instead of horizontal scroll bar is not a problem? In gui files with lots of nested templates, a lot of nodes in the outline view share the same (very long) text prefix. It might be a problem

I have the same fear. Horizontal scrolling will be enabled btw, when the rightmost element reaches 40px, so it won't block them. The icon will be hidden in this case until you scroll. Ideally, I would like to have a TreeTable component,  with the visibility column at the left side of the tree, and free scrolling. I think it can be done, but requires a lot of changes

> * should the on/off switch be recursive? I.e. should turning off go visibility make its sprites invisible? Currently it does not work like that, and I found it puzzling in my tests

Yeah, I think I need to do this. The problem in my opinion though, is that when a node is hidden, it should also automatically block child rendering. Until now, we toggle the visibility of all children. I made a little research to see how similar editors work, and it looks like there is no standard. I guess being able to see a node, without having to make its parent visible has its pros, I just think it's not intuitive. Toggling all children will also be consistent with the existing behavior.

In any case, I won't merge this until we make a decision,

sprocketc committed:
move make-svg-icon-graphic to icons ns

sprocketc committed:
toggle all children

sprocketc committed:
fix styles

sprocketc committed:
style fixes

sprocketc commented:
Recent changes
- We now also toggle the visibility of children. This feels fine, because it depends on the state of a single node, so the expected action will always be triggered.
- Increased the minimum label width to `100px`
- Moved make-svg-icon-graphic to icons ns
- Fixed some color issues introduced here, and also fixed an existing issue with the default/hover color of the icon that was not consistent with the asset tree. I don't know if the specificity changes introduced here #7904 are still relevant, but I tried to maintain them. I would assume `-fx-effect` also has a negative impact, but that requires further investigation.

I can give `TreeTableView` another try , since there is `treeColumnProperty` that can move the visibility column to the left, but the setup feels a little fragile to make additional changes.

sprocketc committed:
Merge branch 'dev' into issue-3916-visibility-toggle

sprocketc committed:
fix event consume

sprocketc committed:
refactor

sprocketc committed:
make visibility toggle stick to the right

sprocketc commented:
I managed to make the toggle button stick to the right by observing multiple values

https://github.com/user-attachments/assets/16d3d320-7ec8-450c-89ce-0f93b0898dd8

I also fixed 
 > - Sometimes, toggling the visibility moves the focus to the first input of the properties panel.
 > - In some cases, the title-bar menu is removed until you click somewhere on the app to trigger a re-render.
 
by adding an additional event filter to stop the propagation. I cannot find a way to get rid of the random flickering that already existed.

sprocketc committed:
Merge branch 'dev' into issue-3916-visibility-toggle

sprocketc committed:
remove unnecessary offset

matgis reviewed (dismissed):
And now it is even better! Love it. ❤️

vlaaad reviewed (dismissed):
This is seriously amazing! Great job!

My only tiny nitpick is that double-clicking the eye icon of a selected tree item opens the referenced file instead of toggling the visibility:

https://github.com/user-attachments/assets/3ef56855-a07e-4ccf-a4fb-4ede2bd585e8

sprocketc committed:
add missing typehint

sprocketc commented:
> My only tiny nitpick is that double-clicking the eye icon of a selected tree item opens the referenced file instead of toggling the visibility:

@vlaaad I wouldn't call that a nitpick :smile: I am aware of this, but I cannot block the action for some reason, even if I add another `MOUSE_CLICKED` filter here https://github.com/defold/defold/pull/10133/files#diff-4c381bb725e8f7d75b6700638fe4d057dbaa9372f3d8cdf7f5bb6c94f104b518R487-R489. This only happens if the element is currently selected. Anyway I pushed a minor fix to add a missing typehint, so I'll need another pass. I appreciate the ultra careful testing btw.

vlaaad reviewed (approved):
LGTM!

selimanac commented:
Great addition! So useful—thank you so much!
But unfortunately, the visibility toggle doesn't do anything on model components. Is this a known issue? Should I open an issue or submit a feature request?

https://github.com/user-attachments/assets/a122973a-2f03-43b1-860a-728978b4d068

selimanac commented:
> @selimanac Hey, thanks for testing this! It looks like there is an open issue #9818.

Ah, thanks! Sorry, I didn't see that.
